File: IsSymmetricTree.java
/**
 * Time Complexity: O(n)
 * Space Complexity: O(1) no extra space
 * Idea:
 * 1. Pass tree as two trees and check if they are mirror of one another. 
 * 2. Gives result if its mirror of itself
 * 
 * Leetcode : Yes
 */
class IsSymmettricTree {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
   }
    public boolean areSymmetricTrees(TreeNode root1, TreeNode root2){
        
        if(root1 == null && root2 == null)
            return true;

        // keys should be same AND
        // left subtree of root1 === right subtree of root2 AND
        // right subtree of root1 === left subtree of root2

        if(root1 != null && root2 != null && root1.val == root2.val && areSymmetricTrees(root1.left, root2.right) && (areSymmetricTrees(root1.right, root2.left))){
            return true;
        }

        return false;
    }

    public boolean isSymmetric(TreeNode root) {
        // pass both trees as root to check if its a mirror of itself
        return areSymmetricTrees(root, root);
    }
    public static void main(String[] args){

        System.out.println("IsSymmettricTree");
        IsSymmettricTree obj = new IsSymmettricTree();
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(2);
        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(4);
        root.right.left = new TreeNode(4);
        root.right.right = new TreeNode(3);

        System.out.println(obj.isSymmetric(root));


    }
}

File: IsSymmetricTree.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Tree
{
    class IsSymmetricTree
    {
        //Time Complexity: O(n)
        //Space Complexity: O(h)
        //Definition of a binary tree node
        public class TreeNode
        {
            internal int val;
            internal TreeNode left;
            internal TreeNode right;
            public TreeNode() { }
            public TreeNode(int val) { this.val = val; }
            public TreeNode(int val, TreeNode left, TreeNode right)
            {
                this.val = val;
                this.left = left;
                this.right = right;
            }
        }
        public bool IsSymmetric(TreeNode root)
        {
            if (root == null)
            {
                return true;
            }
            return helper(root.left, root.right);
        }

        private bool helper(TreeNode left, TreeNode right)
        {
            //base
            if (left == null && right == null)
            {
                return true;
            }
            if (left == null || right == null || left.val != right.val)
            {
                return false;
            }
            return helper(left.left, right.right) && helper(left.right, right.left);
        }

        public bool IsSymmetricIterative(TreeNode root)
        {
            if (root == null)
            {
                return true;
            }
            Stack<TreeNode> s = new Stack<TreeNode>();
            s.Push(root.left);
            s.Push(root.right);
            while (s.Count != 0)
            {
                TreeNode left = s.Pop();
                TreeNode right = s.Pop();
                if (left == null && right == null)
                {
                    continue;
                }
                if (left == null || right == null || left.val != right.val)
                {
                    return false;
                }
                s.Push(left.left);
                s.Push(right.right);
                s.Push(left.right);
                s.Push(right.left);
            }
            return true;
        }


    }
}
